### PR TITLE
if FSA is off, make compilation end analyzer build only analyzer.

### PR DESCRIPTION
**Pull request by heejaechang**
...

this has some issues where now when building,
1 . compilation is needed
2.  compilation end analysis is dynamic not static
3. can cause RPS regression.

### CODE REVIEW

**Comment by heejaechang** [comment]
@jinujoseph I think it is a bit too risky to take for 15.8 and I think it will make build perf to regress.

@mavasani can you take a look? basically, if we want to take this approach, we need to make some change so that we can figure out whether diagnostic analyzer has compilation end analysis without actually initializing analyzer with compilation. otherwise, unfortunately I think this won't work.

File: src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs [hunk 1]
```diff
@@ -511,6 +489,62 @@ public void AddError(ProjectId key, DiagnosticData diagnostic)
...
+                {
+                    return true;
+                }
+
+                return false;
+            }
+
+            private bool SupportedLiveDiagnosticId(Project project, Compilation compilation, string id)
+            {
+                var projectId = project.Id;
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs]: this is only actual meaningful change. all other is just plumbing.

File: src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs [hunk 2]
```diff
@@ -511,6 +489,62 @@ public void AddError(ProjectId key, DiagnosticData diagnostic)
                 AddError(_projectMap, key, diagnostic);
             }
 
+            private bool IsLive(Project project, Compilation compilation, DiagnosticData diagnosticData)
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs]: it used to be a lambda, moved down.

**Comment by heejaechang** [comment]
to see what tests are failing due to this change.

File: src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs [hunk 3]
```diff
@@ -511,6 +489,62 @@ public void AddError(ProjectId key, DiagnosticData diagnostic)
...
+                    }
+
+                    // set ids set
+                    var map = new HashSet<string>();
+                    _liveDiagnosticIdMap.Add(projectId, map);
+
+                    var diagnosticService = _owner._diagnosticService;
+                    foreach (var analyzer in diagnosticService.GetDiagnosticAnalyzers(project))
+                    {
+                        var diagnosticIds = diagnosticService.GetDiagnosticDescriptors(analyzer).Select(d => d.Id);
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs]: this should be inline.

**Comment by sharwell** [comment]
> basically, if we want to take this approach, we need to make some change so that we can figure out whether diagnostic analyzer has compilation end analysis without actually initializing analyzer with compilation. otherwise, unfortunately I think this is too expansive to do while build.

Why not proceed as we already do, and flag analyzers with CEA after `Initialize` is called?

**Comment by heejaechang** [comment]
@sharwell I see. so, wait until analyzer is run and then flag it and remember that. that will work. and if it never ran, then do expansive work. that will work. let me try it.

**Comment by sharwell** [comment]
> and if it never ran, then do expansive work.

This pull request avoids running an analyzer under certain conditions. If the analyzer never ran anyway, there is no need to try to prevent it from running because it already isn't running. 😄

**Comment by CyrusNajmabadi** [comment]
Oooh, interesting.  This says "for personal review" currently.  Would you like more eyes, or would you prefer people wait?  Thanks!

**Comment by heejaechang** [comment]
I am fine with more eyes!

**Comment by CyrusNajmabadi** [comment]
> we need to make some change so that we can figure out whether diagnostic analyzer has compilation end analysis without actually initializing analyzer with compilation. 

Could you explain this bit?  Isn't a compilation relatively cheap to get?  Since everything is deffered in it, it's effectively just a wrapper around a bunch of syntax trees.  I'm not sure i quite understand where the perf concern is wrt "initializing analyzer with compilation."   Is it the cost of running hte analyzer?  Or is it the cost of the compilation itself.

Thanks for any clarification you can provide! :)

**Comment by heejaechang** [comment]
GetCompilationAsync is async since it can be expansive. if everything is already there, then it is cheap, otherwise, compilation tracker needs to bring thing to up to date to return compilation.

creating compilation when everything is there is cheap, since it is not more than object that point to all lazy stuff. but when everything is not ready, it is expansive.

**Comment by CyrusNajmabadi** [comment]
> I am fine with more eyes!

Great.  I'm traveling a bit right now.  But i'll def try to look at thsi when i get a chance.  Def do not wait on me.  But i hope i canhelp out here as this would be *super* useful to have.  

Note: i'm in favor of either two things:
1. Making compilation end actions work (without incurring bad perf).
2. coming up with a separate way that people could do the same sorts of analysis, but in a way we could ensure was fast.

It sounds like this investigation is around camp '1'.  If that works out that would be great as that means no new APIs and existing clients just have a better experience.  However, if it turns out we just can't make it work, then i would still like some sort of solution designed and implemented for these important cases.

Thanks! :)

**Comment by heejaechang** [comment]
oh. Sam answered already :) like I said, this is not long term fix. we are trying to close gap in current implementation around compilation end analyzer.

**Comment by CyrusNajmabadi** [comment]
Oh, sorry, i got a bit confused about things.  If it wasn't too much trouble, it would be nice (for me at least) if the different scenarios were briefly outlined as to their current/upcoming behavior.  i.e. "with FSA off, a compilation-end action currently does X during a build, and Y live.  after this change it will do Z in a build and W afterwards."

I find it a bit hard to keep track of the different permutations of "FSA on/off", "during build or live", "compilation-end-analyzer or not", and were both we are where we're trying to get to (even in hte short term).  Clarity would be appreciated here.

That said, if all of you guys understand it and feel like this is a good step forward, then that's good enough for me as well :)

**Comment by heejaechang** [comment]
design is not that complex. when FSA is on. we run all analyzers live. compilation end analyzer is treated as it used to be. so we are not touching that part.

when FSA is off, right now, compilation end analyzer will not run. since analyzer doesn't run, there is no code fix. problem is when user build, those errors comes in and we say, we can produce live, drop those which is not true when FSA is off. 

this PR is addressing that part. when build gives us errors and if it is compilation end analyzer, we leave mark it as build only errors and don't touch it until next build.

**Comment by etbyrd** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

**Comment by etbyrd** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

File: src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs [hunk 4]
```diff
@@ -308,5 +308,106 @@ private static IEnumerable<AnalyzerPerformanceInfo> Convert(IEnumerable<(Diagnos
         {
             return analyzerPerf.Select(kv => new AnalyzerPerformanceInfo(kv.analyzer.GetAnalyzerId(), DiagnosticAnalyzerLogger.AllowsTelemetry(kv.analyzer, serviceOpt), kv.timeSpan));
         }
+
+        public static bool IsCompilationEndAnalyzer(this DiagnosticAnalyzer analyzer, Project project, Compilation compilation)
+        {
+            // currently, this is only way to see whether analyzer has compilation end analysis or not.
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs]: If we already have the compilation, then why not create a new `CompilationWithAnalyzers` instance with the analyzer and then invoke [GetAnalyzerTelemetryInfoAsync](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/CompilationWithAnalyzers.cs,1153)? This will just force complete all the action registration, and the returned info has the action counts.

Comment (heejaechang) [reply src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs]: looked that API/implementation. it was doing way more work than the simple answer I wanted (it shows by the fact that it is Async method even though CompilationWithAnalyzer already has compilation in it). so, went this pass. do you think it will miss some cases?

example of more works, say analyzer register other hooks, call RunConcurrent or other APIs under initialize, CompilationWithAnalyzer actually do real work for all those calls, which I don't care at all for this question (is compilation end analzyer?), so basically the code does the minimum work to answer the question.

**Comment by etbyrd** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

**Comment by etbyrd** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

**Comment by heejaechang** [comment]
@mavasani can you skim through it?

File: src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs [hunk 5]
```diff
@@ -308,5 +308,113 @@ private static IEnumerable<AnalyzerPerformanceInfo> Convert(IEnumerable<(Diagnos
...
+            public override void RegisterSyntaxNodeAction<TLanguageKindEnum>(Action<SyntaxNodeAnalysisContext> action, ImmutableArray<TLanguageKindEnum> syntaxKinds) { }
+            public override void RegisterSyntaxTreeAction(Action<SyntaxTreeAnalysisContext> action) { }
+            public override void ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags analysisMode) { }
+            public override void EnableConcurrentExecution() { }
+            public override void RegisterOperationAction(Action<OperationAnalysisContext> action, ImmutableArray<OperationKind> operationKinds) { }
+            public override void RegisterOperationBlockAction(Action<OperationBlockAnalysisContext> action) { }
+            public override void RegisterOperationBlockStartAction(Action<OperationBlockStartAnalysisContext> action) { }
+            #endregion
+
+            private class CollectNesteContext : CompilationStartAnalysisContext
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs]: Typo: `CollectNestedContext`

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs [hunk 6]
```diff
@@ -24,30 +24,34 @@ private class StateSet
...
             private readonly VersionStamp _analyzerVersion;
-            private readonly AnalyzerTypeData _analyzerTypeData;
+            private readonly AnalyzerStateNames _analyzerStateNames;
 
             private readonly ConcurrentDictionary<DocumentId, ActiveFileState> _activeFileStates;
             private readonly ConcurrentDictionary<ProjectId, ProjectState> _projectStates;
 
+            // whether this analyzer has compilation end analysis or not
+            // -1 not set, 0 no, 1 yes.
+            private volatile int _compilationEndAnalyzer = -1;
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs]: Consider nullable bool.

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs [hunk 7]
```diff
@@ -255,15 +259,27 @@ public void OnRemoved()
                 InMemoryStorage.DropCache(Analyzer);
             }
 
-            private sealed class AnalyzerTypeData
+            public bool SetCompilationEndAnalyzer(Project project, Compilation compilation)
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs]: Split into two methods `bool GetIsCompilationEndAnalyzer` and `void ComputeIsCompilationEndAnalyzer`

File: src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs [hunk 8]
```diff
@@ -398,24 +373,24 @@ private class InprogressState
             private int _incrementDoNotAccessDirectly = 0;
 
             private readonly ExternalErrorDiagnosticUpdateSource _owner;
-            private readonly Solution _solution;
-
             private readonly HashSet<ProjectId> _builtProjects = new HashSet<ProjectId>();
-            private readonly Dictionary<ProjectId, HashSet<string>> _diagnosticIdMap = new Dictionary<ProjectId, HashSet<string>>();
+
+            private readonly Dictionary<ProjectId, HashSet<string>> _allDiagnosticIdMap = new Dictionary<ProjectId, HashSet<string>>();
+            private readonly Dictionary<ProjectId, HashSet<string>> _liveDiagnosticIdMap = new Dictionary<ProjectId, HashSet<string>>();
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs]: Consider `Dictionary<ProjectId, ImmutableHashSet<string>>`

File: src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs [hunk 9]
```diff
@@ -511,6 +491,62 @@ public void AddError(ProjectId key, DiagnosticData diagnostic)
...
+                    }
+
+                    var fullSolutionAnalysis = ServiceFeatureOnOffOptions.IsClosedFileDiagnosticsEnabled(project);
+                    if (!project.SupportsCompilation || fullSolutionAnalysis)
+                    {
+                        return SupportedDiagnosticId(project.Id, id);
+                    }
+
+                    // set ids set
+                    var map = new HashSet<string>();
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs]: Rename to `set`?

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs [hunk 10]
```diff
@@ -24,7 +24,7 @@ private class StateSet
 
             // analyzer version this state belong to
             private readonly VersionStamp _analyzerVersion;
-            private readonly AnalyzerTypeData _analyzerTypeData;
+            private readonly AnalyzerStateNames _analyzerStateNames;
```
Comment (sharwell) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs]: ❓ How is this change related to _this_ pull request?

Comment (heejaechang) [reply src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs]: while working on this area. saw and changed. 

first thought putting the bool in this AnalyzerTypeData, but decide not to for now, since IsCompilationEndAnalyzer require compilation which is different per project.

**Comment by mavasani** [comment]
Ping @heejaechang. Anything blocking this?

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs (hunk 6 above)
Comment (heejaechang) [reply src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs]: used int so that I can make it volatile.

**Comment by heejaechang** [comment]
@jinujoseph @mavasani do we still want this for dev16.0 p1? or master is fine?

**Comment by jinujoseph** [comment]
Master is fine.

**Comment by heejaechang** [comment]
retargeted to master

**Comment by heejaechang** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by heejaechang** [comment]
retest ubuntu_16_mono_debug_prtest please

**Comment by heejaechang** [comment]
retest ubuntu_16_mono_debug_prtest please

**Comment by heejaechang** [comment]
looks like VS integration tests are busted. by looking at the result, it looks like it is expecting "DEBUG", but the active configuration is "Release". 

there is VS bug on this issue (active configuration automatically becomes "release") on platform side which got fixed recently? is this failure related to that issue?

tagging @dotnet/roslyn-infrastructure to see whether this is a known issue?

**Comment by heejaechang** [comment]
close and reopen for roslyn-CI to re-run
